Fix for CR-1097956: xbutil2 validate shows success for 2RP platform when PLP not loaded

vboggara-xilinx opened the pull request:
<!-- Thanks for sending a pull request! Please fill out below, remove sections that don't apply for your pull request.  -->
#### Problem solved by the commit
https://jira.xilinx.com/browse/CR-1097956
#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Test results for validate were unclear for users in below cases
1. when tests are skipped due to the shell platform not being programmed.
2. when running a single test in the event of a skipped test, 
as they would see "Validation Completed" with an exit code of 0 without any additional information about the test.
#### How problem was solved, alternative solutions (if any) and why they were rejected
The validate result for a single test or all test will now always be reported, regardless if the test is skipped/failed/passed. A non-zero exit code will be returned and "Validation Failed" will be displayed for the skipped case now (in addition to the failed case), 
"Validation Failed" will be displayed for users in below scenarios.
1. when all tests or single test skipped except default tests like aux-connection, pcie-link, etc.
2.  When all tests or single test failed.
Reference PR: https://github.com/Xilinx/XRT/pull/6853
#### Risks (if any) associated the changes in the commit
This CR changes what the user can see when running ''xbutil validate
#### What has been tested and how, request additional testing if necessary
Single test runs and multi-test runs were tested.
Example of skipped test.
`

xbutil validate -d
Validate Device           : [0000:86:00.1]
    Platform              : xilinx_u250_gen3x16_base_4
    SC Version            : 4.6.20
    Platform ID           : F8DAC62E-49D9-B0AA-E9FC-6F260D9D0DFB

Test 1 [0000:86:00.1]     : aux-connection
    Test Status           : [PASSED]

Test 2 [0000:86:00.1]     : pcie-link
    Test Status           : [PASSED]

Test 3 [0000:86:00.1]     : sc-version
    Test Status           : [PASSED]

Test 4 [0000:86:00.1]     : verify
Test 5 [0000:86:00.1]     : dma
Test 6 [0000:86:00.1]     : iops                                                
Test 7 [0000:86:00.1]     : mem-bw
Test 8 [0000:86:00.1]     : p2p
Test 9 [0000:86:00.1]     : vcu
Test 10 [0000:86:00.1]    : aie-pl
Validation failed. Please run the command --verbose option for more detail

xbutil validate -d
Validate Device           : [0000:86:00.1]
    Platform              : xilinx_u250_gen3x16_xdma_shell_4_1
    SC Version            : 4.6.20
    Platform ID           : 12C8FAFB-0632-499D-B1C0-C6676271B8A6

Test 1 [0000:86:00.1]     : aux-connection 
    Test Status           : [PASSED]

Test 2 [0000:86:00.1]     : pcie-link 
    Test Status           : [PASSED]

Test 3 [0000:86:00.1]     : sc-version 
    Test Status           : [PASSED]

Test 4 [0000:86:00.1]     : verify 
    Test Status           : [PASSED]                                            

Test 5 [0000:86:00.1]     : dma 
    Details               : Buffer size - '16 MB'                               
                            Host -> PCIe -> FPGA write bandwidth = 9041.7 MB/s
                            Host <- PCIe <- FPGA read bandwidth = 11944.3 MB/s
                            Buffer size - '16 MB'
                            Host -> PCIe -> FPGA write bandwidth = 9112.1 MB/s
                            Host <- PCIe <- FPGA read bandwidth = 11918.5 MB/s
                            Buffer size - '16 MB'
                            Host -> PCIe -> FPGA write bandwidth = 9084.4 MB/s
                            Host <- PCIe <- FPGA read bandwidth = 11914.5 MB/s
                            Buffer size - '16 MB'
                            Host -> PCIe -> FPGA write bandwidth = 8962.5 MB/s
                            Host <- PCIe <- FPGA read bandwidth = 11918.1 MB/s
    Test Status           : [PASSED]

Test 6 [0000:86:00.1]     : iops 
    Details               : IOPS: 544268 (verify)                               
    Test Status           : [PASSED]

Test 7 [0000:86:00.1]     : mem-bw 
    Error(s)              : Found Platform                                      
                            Platform Name: Xilinx
                            INFO: Reading
                            /opt/xilinx/firmware/u250/gen3x16/xdma-shell/test/bandwidth.xclbin
                            
    Test Status           : [FAILED]

Validation failed. Please run the command --verbose option for more details
`

#### Documentation impact (if any)
none.

gbuildx commented:
Build Failed! :(

rozumx reviewed (changes requested):
With the exception of 1 location, this code is basically doing enabling "verbosity" if there is only 1 validation test.   To address this "verbosity" requirement, a new variable was introduced and passed to all the descriptions to indicate there is only one test.   

Unfortunately, this type of solution is something we should avoid for "business rule logic" is being mixed in with functionality.   This type of solution will lead to scalability issues; in other words, the code will be difficult to support.

Instead, the following should be done:
1. In SubCmdValidate::print_status(), the following business rule logic" should be added at around line 1515:
- If there is only 1 test to be executed, then the verbosity flag should automatically be set.
- If the verbosity flag is already set, then do nothing.
- If the verbosity flag is not set, then set it along with producing a message indicating that since only 1 test is being executed that the verbosity flag is being enabled.

2. Refactor the code logic for SubCmdValidate::pretty_print_test_desc() to NOT use a "single_case" variable.  In fact, once could argue that on line 1358 instead of using "single_case" one should use XBU::getVerbose().

dayeh-xilinx commented:
retest this please.

rchane commented on src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:
```diff
@@ -1351,16 +1351,16 @@ get_test_name(const std::string& input_name)
 static void
 pretty_print_test_desc(const boost::property_tree::ptree& test, int& test_idx,
-                       std::ostream & _ostream, const std::string& bdf)
+                       std::ostream & _ostream, const std::string& bdf, bool single_case)
```
As Stephen mentioned, the usage of the single_case variable here isn't ideal and contributed to why I closed the initial PR I made. I adhere to his review for his proposed changes.

vboggara-xilinx replied:
Addressed Stephen's comments and submitted changes in PR

gbuildx commented:
Build Failed! :(

dayeh-xilinx commented:
retest this please.

gbuildx commented:
Build Failed! :(

dayeh-xilinx commented:
retest this please.

gbuildx commented:
Build Failed! :(

chvamshi-xilinx commented:
> With the exception of 1 location, this code is basically doing enabling "verbosity" if there is only 1 validation test. To address this "verbosity" requirement, a new variable was introduced and passed to all the descriptions to indicate there is only one test.
> 
> Unfortunately, this type of solution is something we should avoid for "business rule logic" is being mixed in with functionality. This type of solution will lead to scalability issues; in other words, the code will be difficult to support.
> 
> Instead, the following should be done:
> 
> 1. In SubCmdValidate::print_status(), the following business rule logic" should be added at around line 1515:
> 
> * If there is only 1 test to be executed, then the verbosity flag should automatically be set.
> * If the verbosity flag is already set, then do nothing.
> * If the verbosity flag is not set, then set it along with producing a message indicating that since only 1 test is being executed that the verbosity flag is being enabled.
> 
> 2. Refactor the code logic for SubCmdValidate::pretty_print_test_desc() to NOT use a "single_case" variable.  In fact, once could argue that on line 1358 instead of using "single_case" one should use XBU::getVerbose().

This is very good suggestion. Thanks Stephen.

vboggara-xilinx commented:
> With the exception of 1 location, this code is basically doing enabling "verbosity" if there is only 1 validation test. To address this "verbosity" requirement, a new variable was introduced and passed to all the descriptions to indicate there is only one test.
> 
> Unfortunately, this type of solution is something we should avoid for "business rule logic" is being mixed in with functionality. This type of solution will lead to scalability issues; in other words, the code will be difficult to support.
> 
> Instead, the following should be done:
> 
> 1. In SubCmdValidate::print_status(), the following business rule logic" should be added at around line 1515:
> 
> * If there is only 1 test to be executed, then the verbosity flag should automatically be set.
> * If the verbosity flag is already set, then do nothing.
> * If the verbosity flag is not set, then set it along with producing a message indicating that since only 1 test is being executed that the verbosity flag is being enabled.
> 
> 2. Refactor the code logic for SubCmdValidate::pretty_print_test_desc() to NOT use a "single_case" variable.  In fact, once could argue that on line 1358 instead of using "single_case" one should use XBU::getVerbose().

Thanks Stephen, addressed your comments and submitted the PR,

vboggara-xilinx commented:
Below is the Run log with the latest changes.

Run1:
------
root@rahul:/home/rahul/naresh/XRT/XRT/build# xbutil validate -d
WARNING: Unexpected xocl version (2.14.257) was found. Expected 2.14.0, to match XRT tools.
Validate Device           : [0000:86:00.1]
    Platform              : xilinx_u250_gen3x16_base_4
    SC Version            : 4.6.20
    Platform ID           : F8DAC62E-49D9-B0AA-E9FC-6F260D9D0DFB
-------------------------------------------------------------------------------
Test 1 [0000:86:00.1]     : aux-connection 
    Test Status           : [PASSED]
-------------------------------------------------------------------------------
Test 2 [0000:86:00.1]     : pcie-link 
    Test Status           : [PASSED]
-------------------------------------------------------------------------------
Test 3 [0000:86:00.1]     : sc-version 
    Test Status           : [PASSED]
-------------------------------------------------------------------------------
Test 4 [0000:86:00.1]     : verify 
Test 5 [0000:86:00.1]     : dma 
Test 6 [0000:86:00.1]     : iops                                                
Test 7 [0000:86:00.1]     : mem-bw 
Test 8 [0000:86:00.1]     : p2p 
Test 9 [0000:86:00.1]     : vcu 
Test 10 [0000:86:00.1]    : aie-pl 
Validation failed. Please run the command '--verbose' option for more details

Run2:
------
root@rahul:/home/rahul/naresh/XRT/XRT/build# xbutil validate -r vcu -d
WARNING: Unexpected xocl version (2.14.257) was found. Expected 2.14.0, to match XRT tools.
Validate Device           : [0000:86:00.1]
    Platform              : xilinx_u250_gen3x16_base_4
    SC Version            : 4.6.20
    Platform ID           : F8DAC62E-49D9-B0AA-E9FC-6F260D9D0DFB
-------------------------------------------------------------------------------
Verbose: Enabling Verbosity
Test 1 [0000:86:00.1]     : vcu 
    Description           : Run decoder test
    Details               : Verify xclbin not available or shell partition is not
                            programmed. Skipping validation.
    Test Status           : [SKIPPED]
-------------------------------------------------------------------------------
Validation failed

Run3:
------
root@rahul:/home/rahul/naresh/XRT/XRT/build# xbutil validate -r vcu aie-pl -d
WARNING: Unexpected xocl version (2.14.257) was found. Expected 2.14.0, to match XRT tools.
Validate Device           : [0000:86:00.1]
    Platform              : xilinx_u250_gen3x16_base_4
    SC Version            : 4.6.20
    Platform ID           : F8DAC62E-49D9-B0AA-E9FC-6F260D9D0DFB
-------------------------------------------------------------------------------
Test 1 [0000:86:00.1]     : vcu 
Test 2 [0000:86:00.1]     : aie-pl 
Validation failed. Please run the command '--verbose' option for more details

gbuildx commented:
Build Failed! :(

chvamshi-xilinx commented on src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:
```diff
@@ -1534,12 +1540,19 @@ run_test_suite_device( const std::shared_ptr<xrt_core::device>& device,
 
     pretty_print_test_run(ptTest, status, std::cout);
 
+    // consider only when testcase is part of lack_box_tests.
```
typo in the comment

vboggara-xilinx replied:
addressed.

chvamshi-xilinx commented on src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:
```diff
@@ -1510,21 +1510,29 @@ run_test_suite_device( const std::shared_ptr<xrt_core::device>& device,
     // we need to print the test description before running the test
     auto is_black_box_test = [ptTest]() {
-      std::vector<std::string> black_box_tests = {"verify", "mem-bw", "iops", "vcu", "aie-pl", "dma", "p2p"};
+      std::vector<std::string> black_box_tests = {"verify", "mem-bw", "iops", "vcu", "aie-pl", "dma", "p2p"};   
```
Added an extra space at the end.

vboggara-xilinx replied:
addressed.

chvamshi-xilinx reviewed (approved):
Remove extra spaces at end of the line.
Other changes are looking good.

chvamshi-xilinx commented:
retest this please

gbuildx commented:
Build Failed! :(

manikandan-xilinx commented:
retest this please.

gbuildx commented:
Build Failed! :(

manikandan-xilinx commented:
retest this please.

gbuildx commented:
Build Failed! :(

chvamshi-xilinx commented:
retest this please

gbuildx commented:
Build Failed! :(

dayeh-xilinx commented:
retest this please.

gbuildx commented:
Build Failed! :(

vboggara-xilinx commented:
adding do not merge tag because board farm infra required some changes in running xbutil validate.

chvamshi-xilinx commented:
board farm infra changes are done. Retesting this

chvamshi-xilinx commented:
retest this please

gbuildx commented:
Build Passed!

vboggara-xilinx commented:
@rozumx Please review this PR.
